### PR TITLE
[WIP] Add password_fill userscript

### PR DESCRIPTION
**Pull request by t-wissmann**
Add a configurable userscript that fills login forms (i.e. the fiels
"Username" and "Password) of websites using a configurable backend where
the actual passwords are stored. The only backend yet is using the
password store "pass".

[Review on Reviewable](https://reviewable.io/reviews/the-compiler/qutebrowser/873)

### CODE REVIEW

**Comment by lamarpavel** [comment]
Security concern: If I see that correctly, this script immediately pastes username and password into fields marked as "password" and it is randomly selecting possible matches from your .password-store. It would be fatal to paste into the wrong field, such as a search providing auto-complete via ajax etc.

My suggestion would be to do a hint-selection first where all potential password-fields are highlighted and only after selecting a hint would the password be pasted.

Aside from that, it's nice to see integration for `pass` spreading. I like the idea of including prompts like dmenu, rofi etc. into the script.

**Comment by t-wissmann** [comment]
The behaviour is:

1. Pick matching password entry based on URL (configurable)
2. Choose one of them by random or menu (depending on configuration)
3. Paste that into every form that has at least one password field, (i.e. some `input` tag with `type=password`).

The selection of possible matches from the password-store is not completely random. And furthermore you can configure `password_fill` how to choose these passwords (only by filename or by "URL:"-line). If you still fear, that the wrong password, entry is connected, just customize `choose_entry` to use some menu.

But beside choosing the wrong entry in `choose_entry`, I don't share your concern. Consider that you are on page *X.Y* and you want to login using your credentials (username and your personal password for page *X.Y*). Then there is no problem at all, filling your credentials into any form that has a password field, because even if you fill it into only the correct field, the adminstrator of *X.Y* has the full power over your credentials. So there is no need for him to trick you using some Ajax-Search-Form-With-Password-Field.

TL;DR: If the adminstrator of some site want's to publish your credentials somewhere, he can do it already now without you using a password filler.

That's the reason why one does not need "Hint selection for the correct login form". Furthermore, it is very hard or not possible at all to implement that as a userscript.

**Comment by lamarpavel** [comment]
>TL;DR: If the adminstrator of some site want's to publish your credentials somewhere, he can do it already now without you using a password filler.

Unless the site is seriously misconfigured, the administrator will not have access to your password. But that is beside the point. I was thinking along the lines of mistakes made by the user, eg you are on google.com and accidentally press the wrong keys and something gets pasted into the search bar, for example (shouldn't happen there because the google search bar is probably not tagges as `type=password`, but you get the idea).

However, when configured to spawn a prompt like passmenu for the selection, that should already be sufficient if you can abort with ESC and nothing gets sent. That would at least rule out the immediate consequences of an accidental key-press.

**Comment by t-wissmann** [comment]
> > TL;DR: If the adminstrator of some site want's to publish your credentials somewhere, he can do it already now without you using a password filler.

> Unless the site is seriously misconfigured, the administrator will not have access to your password.

Each time you use your password, it is sent in plaintext to the adminstrator, and then the admin decides what to do with it. And that's why it doesn't make a difference in which login form you paste your password as long as you trust the administrator.

> eg you are on google.com and accidentally press the wrong keys and something gets pasted into the search bar, (shoudn't happen because probably not tagges as `type=password`)

That's the first reason why it won't happen. The second is that it will only paste passwords which contain `google.com` in the filename or a line containing `URL: google.com` (both configurable). So the worst thing that could happen is that you are googling for your google password.

And as noticed by you, you can always configure it to use a menu for confirmation.

**Comment by lamarpavel** [comment]
>Each time you use your password, it is sent in plaintext to the adminstrator, and then the admin decides what to do with it.

I really hope you don't enter passwords on sites whose connection isn't secured by TLS. Often, the password is hashed and than only the hash is compared to the hash in the database, but even if that weren't the case, the administrator would have to sniff the connection, missusing the sites certificate. Which is what I meant by the site being misconfigured. Anyway I don't mean to derail your PR any further. Thanks for this userscript (:

File: misc/userscripts/password_fill [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+fi
+
+error() {
+    local msg="$*"
+    echo "message-error '${msg//\'/\\\'}'" >> "$QUTE_FIFO"
+}
+msg() {
+    local msg="$*"
+    echo "message-info '${msg//\'/\\\'}'" >> "$QUTE_FIFO"
+}
```
Comment (The-Compiler) [inline misc/userscripts/password_fill]: Hmm - maybe `:message-error` and `:message-info` should be `maxsplit=0` commands so they interpret everything literally and quoting like this is unneeded?

File: misc/userscripts/password_fill [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+    exit 0
+}
+
+# ======================================================= #
+# CONFIGURATION
+# ======================================================= #
+# The configuration file is per default located in
+# ~/.config/qutebrowser/password_fill_rc and is a bash script that is loaded
+# later in the present script. So basically you can replace all of the
+# following definitions and make them fit your needs.
```
Comment (The-Compiler) [inline misc/userscripts/password_fill]: That's a nice idea!

File: misc/userscripts/password_fill [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+match_line_pattern='^url: .*' # applied using grep -iE
+user_pattern='^(user|username): '
+
+GPG_OPTS=( "--quiet" "--yes" "--compress-algo=none" "--no-encrypt-to" )
+GPG="gpg"
+export GPG_TTY="${GPG_TTY:-$(tty 2>/dev/null)}"
+which gpg2 &>/dev/null && GPG="gpg2"
+[[ -n $GPG_AGENT_INFO || $GPG == "gpg2" ]] && GPG_OPTS+=( "--batch" "--use-agent" )
+
+pass_backend() {
```
Comment (The-Compiler) [inline misc/userscripts/password_fill]: I'm unaware with how `pass` works - but this doesn' call `pass`, right? Does it just use gpg to read `pass` files?

File: misc/userscripts/password_fill [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+        done < <($GPG "${GPG_OPTS}" -d "$path" )
+    }
+}
+# =======================================================
+
+# load some sane default backend
+reset_backend
+pass_backend
+# load configuration
+QUTE_CONFIG_DIR=${QUTE_CONFIG_DIR:-${XDG_CONFIG_HOME:-$HOME/.config}/qutebrowser/}
```
Comment (The-Compiler) [inline misc/userscripts/password_fill]: Adding `QUTE_CONFIG_DIR` (and some other `*_DIR` things) to the userscript variables for qutebrowser sounds like a good idea.

File: misc/userscripts/password_fill [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+match_filename=1 # whether allowing entry match by filepath
+match_line=1     # whether allowing entry match by URL-Pattern in file
+match_line_pattern='^url: .*' # applied using grep -iE
+user_pattern='^(user|username): '
+
+GPG_OPTS=( "--quiet" "--yes" "--compress-algo=none" "--no-encrypt-to" )
+GPG="gpg"
+export GPG_TTY="${GPG_TTY:-$(tty 2>/dev/null)}"
+which gpg2 &>/dev/null && GPG="gpg2"
+[[ -n $GPG_AGENT_INFO || $GPG == "gpg2" ]] && GPG_OPTS+=( "--batch" "--use-agent" )
```
Comment (The-Compiler) [inline misc/userscripts/password_fill]: Should those really be upper-case, seeing that the user can't override them by using environment variables anyways?

**Comment by t-wissmann** [comment]
> > Each time you use your password, it is sent in plaintext to the adminstrator, and then the admin decides what to do with it.

> I really hope you don't enter passwords on sites whose connection isn't secured by TLS. Often, the password is hashed and than only the hash is compared to the hash in the database, but even if that weren't the case, the administrator would have to sniff the connection, missusing the sites certificate. Which is what I meant by the site being misconfigured.

Though the password is sent over a TLS-secured connection, the administrator will receive it in plain text! Even if the password is eventually hashed for database lookup, the administrator receives the password in plain text. That means: Even if you do TLS and password-hashing in your database, each time you login, you present the password to the administrator in plaintext. And that's why it's not that bad if you hit the wrong form on a site, as long as you trust the administrator.

> Anyway I don't mean to derail your PR any further. Thanks for this userscript (:

Thanks for your feedback!

File: misc/userscripts/password_fill (hunk 1 above)
Comment (t-wissmann) [inline misc/userscripts/password_fill]: Not sure what that precisely means, but I imagine that one wants to bind something like: "message-info launching terminal ; spawn terminal ; message-info terminal launched!". Would `maxsplit=0` imply that the second and third command are interpreted als parameters to the first `mesage-info`? Alternatively, we could do something like $(qutebrowser --shell-escape < "$msg") that escapes all the critical characters away. (similar to in some programming languages?)

File: misc/userscripts/password_fill (hunk 3 above)
Comment (t-wissmann) [inline misc/userscripts/password_fill]: Right. I directly access passes backend, because the output of the `pass` command is not nicely parsable for scripts (especially that of `pass grep`).

File: misc/userscripts/password_fill (hunk 5 above)
Comment (t-wissmann) [inline misc/userscripts/password_fill]: Good question. The user can only override them in the config file. I copied these GPG-variables from `pass`: see `vim $(which pass)`. I hoped they are chosen wisely. Of course I could/should make `GPG` overridable, but I don't see a way to do that for `GPG_OPTS` as it is an array. Do you want me to make `GPG_OPTS` lower case and `GPG` overidable?

**Comment by The-Compiler** [comment]
@t-wissmann IIRC you told me you don't consider this ready yet - any update so far?

**Comment by t-wissmann** [comment]
I think, now it's ready :-)

**Comment by torsava** [comment]
Forgive my impatience, but when could I expect this to be merged? Are there still some problems that need to be solved?

**Comment by The-Compiler** [comment]
@torsava I'm sorry for not merging this yet, it's just me getting lost in too many pull requests :wink: I'll need to open a few issues for the qutebrowser-feature-requests I noticed and commented on this PR, and then I'll merge this. If it's not merged by the end of the week, feel free to ping me again :wink:

**Comment by torsava** [comment]
Ok, big thanks for your time!

**Comment by tex** [comment]
Hi @The-Compiler ,ping. I'm eagerly waiting on this feature, using pass a lot and this is the only thing I'm still using Firefox for. How does it look with this feature?

**Comment by The-Compiler** [comment]
@tex huh? This was merged half a year ago :wink:

**Comment by tex** [comment]
oops, sorry, I had to overlook it somehow :-)

**Comment by tex** [comment]
I tried to use it, read the doc, but I was unable to get it working.  Can anyone using it create some howto  (screencast) showing how to configure it and use it please?

**Comment by The-Compiler** [comment]
Where are you stuck? I've never used it myself though :wink:

**Comment by tex** [comment]
Well, I copied password_fill script to ~/.config/qutebrowser and then start qutebrowser, there I executed `spawn --userscript ~/.config/qutebrowser/password_fill` as it is what it tells in its help, and nothing happens. I don't know how it should work, what it should do, should I press something, how to make it work? I use pass, for my passwords, and for my Firefox passwords - I use pass-manager addons there - and it asks for password at begining, and when I visit some site where I enter password it automatically saves that password, when I enter site again, username+password is automatically filled. I see nothing like that in qutebrowser and password_fill script.

**Comment by The-Compiler** [comment]
FWIW @claudehohl mentioned he wants something like this as well, and when I told him about it, he said he might be interested in writing a tutorial for it... so, ping? :wink:

**Comment by wichtounet** [comment]
Is there any support for htaccess-like completion ?

**Comment by The-Compiler** [comment]
@wichtounet There's support for `~/.netrc` [files](https://linux.die.net/man/5/netrc) for HTTP basic authentication, but it's undocumented currently, and you'll need to store your credentials in plaintext.

**Comment by t-wissmann** [comment]
@tex : Does your pass entries have the correct format?
```
Pass backend: (see also passwordstore.org)
  This script expects pass to store the credentials of each page in an extra
  file, where the filename (or filepath) contains the domain of the respective
  page. The first line of the file must contain the password, the login name
  must be contained in a later line beginning with "user:", "login:", or
  "username:" (configurable by the user_pattern variable).
```

The expected behaviour is: run this script on some website. If there is some login-form, then the script fills in the website's credentials from your password store.

If you use rofi, I strongly recommend the following `password_fill_rc`:
```
# Show all password fields in the menu
query_entries() {
    # safe queried url for choose_entry
    export queried_url="$1"
    mapfile -t files < <(find -L "$PREFIX" -iname '*.gpg' -printf '%P\n' |sed 's,\.gpg$,,')
}

# Even if there is only one entry, always show a menu
# for user confirmation.
choose_entry() {
    MENU_COMMAND=(
        rofi -dmenu
            -p "qutebrowser> "
            -filter "$queried_url"
            -mesg $'Pick a password entry for <b>'"${QUTE_URL//&/&amp;}"'</b>'
    )
    file=$( printf "%s\n" "${files[@]}" | "${MENU_COMMAND[@]}" )
}
```

**Comment by yorkbai** [comment]
I try this plugin almost same as @tex 

1. cp password-fill to  ~/Library/Application\ Support/qutebrowser/userscripts
2. add to pass in commandline(#pass edit web/x.com) , 'll create file ~/.password-store/web/x.gpg 
3. open qutebrowser and call " :spawn --userscript password_fill"  , it's said "Username not set in entry web/x.com"

 x.gpg only two line like following:
     *****            # it's password of login_user
     user:  login_user

sth I missed ? plz !

**Comment by t-wissmann** [comment]
Is the `user: login_user` at the beginning of the line? There must be no whitespace before! Other than that, I don't see why you are having this issue, when comparing your description with the password_fill source code.

**Comment by yorkbai** [comment]
Yes, There no whitespace . I checking the login page like this.
 
login

**Comment by t-wissmann** [comment]
Actually, I was talking about the content of your `web/x.com.gpg` file. Does password fill work for you on any other website?

**Comment by yorkbai** [comment]
I try it it in little other website,  havn't working normal  also.

**Comment by t-wissmann** [comment]
I'm sorry but I don't have any further idea how to help you.
